GVFS/GVFS/CommandLine/StatisticsVerb.cs

```diff
@@ -0,0 +1,209 @@
+﻿using CommandLine;
+using GVFS.Common;
+using GVFS.Common.Database;
+using GVFS.Common.FileSystem;
+using GVFS.Common.Git;
+using GVFS.Common.NamedPipes;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace GVFS.CommandLine
+{
+    [Verb(StatisticsVerb.StatisticsVerbName, HelpText = "Get statistics for the health state of the repository")]
+    public class StatisticsVerb : GVFSVerb.ForExistingEnlistment
+    {
+        private const string StatisticsVerbName = "statistics";
+
+        private const int DirectoryDisplayCount = 5;
+
+        [Option(
+            'd',
+            "directory",
+            Required = false,
+            HelpText = "Run the statistics tool on a specific directory")]
+        public string Directory { get; set; }
+
+        protected override string VerbName
+        {
+            get { return StatisticsVerbName; }
+        }
+
+        protected override void Execute(GVFSEnlistment enlistment)
+        {
+            // The path to the root of git's tree
+            string sourceRoot = enlistment.DotGitRoot.Substring(0, enlistment.DotGitRoot.Length - 4);
+
+            if (this.Directory == null)
+            {
+                this.Directory = string.Empty;
+            }
+            else if (this.Directory.Equals("."))
+            {
+                this.Directory = Environment.CurrentDirectory.Substring(sourceRoot.Length);
+            }
+
+            // Get all of the data needed to calculate statistics then pass them into GVFSEnlistmentStatistics
+            List<string> modifiedPathsFileList = new List<string>();
+            List<string> modifiedPathsFolderList = new List<string>();
+            List<string> placeholderFilePathList = new List<string>();
+            List<string> placeholderFolderPathList = new List<string>();
+
+            this.GetPlaceholdersFromDatabase(enlistment, out List<IPlaceholderData> filePlaceholders, out List<IPlaceholderData> folderPlaceholders);
+            foreach (IPlaceholderData placeholderData in filePlaceholders)
+            {
+                placeholderFilePathList.Add(placeholderData.Path);
+            }
+
+            foreach (IPlaceholderData placeholderData in folderPlaceholders)
+            {
+                placeholderFolderPathList.Add(placeholderData.Path);
+            }
+
+            foreach (string path in this.GetModifiedPathsFromPipe(enlistment))
+            {
+                if (path.Last() == GVFSConstants.GitPathSeparator)
+                {
+                    path.TrimEnd('/');
+                }
+                else
+                {
+                    modifiedPathsFileList.Add(path);
+                }

```

Comment:
Should this code block be:
```
if (path.Last() != GVFSConstants.GitPathSeparator)
{
    modifiedPathsFileList.Add(path);
}
```
Since the `path.TrimEnd('/')` doesn't affect anything

Does this mean that we aren't counting any directory entries in the ModifiedPaths list?  Oh wait maybe you simply missed adding to `modifiedPathsFolderList`?

This would be another reason to be using the `git ls-files -sv` and count the entries with the skipworktree bit set.  Then we would have all the files that are affected by the ModifiedPaths.

Reply:
You're so right, I totally just whiffed and forgot to call `modifiedPathsFolderList.Add(path)`. I'm a little confused about the skip worktree bit and how it would change what I'm doing. Am I missing a critical statistic since I'm never looking at it? I'm just legitimately unsure where it comes into all of this

Reply:
Maybe I am missing something but if I have the following in my modified paths file:
```
.gitattributes
foo/
```

And there are 1 million files under the `foo` directory and it is the only directory in the repo what would the statistics for modified paths be?  It should be `100%`.  But for the same repository that has a million files in the `foo` directory and I have `bar/` in my modified paths or 1 million paths that are not under `foo` the percent that git was keeping up to date would be very small close to (`0%`) even though I have a million entries in the modified paths.  These entries could still affect performance if we have a `bar/` in the modified paths and it is a folder with a million files but they are have not been added to git for tracking, you would have the close to `0%` but performance would still be bad because `git status` would need to go through the directories of `bar/` to find the untracked files.

Using `git ls-files -sv` you would easily be able to see that git has 1 million files in the index and all 1 million have the skip-worktree bit off in the case of `foo/` or only one has it off in the case of `bar` or whatever percentage of all the files that are in the index.  You would also be able to do some checks on the placeholder entries because the files in the index with the skip-worktree bit on are the ones that can be placeholders and you could get stats on how many of those entries placeholders have been created for.

So you have:
```
# index entries (All the files git is tracking)
# index entries with skip-worktree bit (these could be placeholders and affect the performance of updating placeholders)
# index entries without skip-worktree bit (This bit is turned off by entries in the modified paths list and these are what git is in control of keeping up to date and will affect git command performance)
```

These numbers could be very different from the number of modified paths.  Which is why I'm wondering if we should pay attention to it.

Hope this make some sort of sense, let me know.